compiler/noirc_evaluator/src/ssa/opt/remove_unreachable_instructions.rs

```diff
@@ -0,0 +1,465 @@
+//! This module defines an SSA pass to remove instructions that are unreachable.
+//! For example, if an instruction in a block is `constrain u1 0 == u1 1`,
+//! any subsequent instructions in that block will never be executed. This pass
+//! then removes those subsequent instructions and replaces the block's terminator
+//! with a special `unreachable` value.
+
+use crate::ssa::{
+    ir::{
+        dfg::DataFlowGraph,
+        function::Function,
+        instruction::{Instruction, TerminatorInstruction},
+    },
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    pub(crate) fn remove_unreachable_instructions(mut self) -> Ssa {
+        for function in self.functions.values_mut() {
+            function.remove_unreachable_instructions();
+        }
+        self
+    }
+}
+
+impl Function {
+    fn remove_unreachable_instructions(&mut self) {
+        // The current block we are currently processing
+        let mut current_block_id = None;
+
+        // Whether the current block instructions were determined to be unreachable
+        // after an always failing one was found.
+        let mut current_block_instructions_are_unreachable = false;
+
+        self.simple_reachable_blocks_optimization(|context| {
+            let block_id = context.block_id;
+
+            if current_block_id != Some(block_id) {
+                current_block_id = Some(block_id);
+                current_block_instructions_are_unreachable = false;
+            }
+
+            if current_block_instructions_are_unreachable {
+                context.remove_current_instruction();
+                return;
+            }
+
+            if always_fails(context.dfg, context.instruction()) {
+                current_block_instructions_are_unreachable = true;
+
+                let terminator = context.dfg[block_id].take_terminator();
+                let call_stack = terminator.call_stack();
+                context.dfg[block_id]
+                    .set_terminator(TerminatorInstruction::Unreachable { call_stack });
+            }
+        });
+    }
+}
+
+/// Returns `true` if the given instruction will always produce an asertion failure.
+fn always_fails(dfg: &DataFlowGraph, instruction: &Instruction) -> bool {
```

Comment:
Nice - we can extend this in the future to things like array get/sets with indices which are known to overflow

Reply:
True! I was thinking a range check could also always fail, but I didn't include it in this PR for simplicity.